dsp/modules/tensorrt_llm.py

```diff
@@ -0,0 +1,229 @@
+from pathlib import Path
+from typing import Any, Optional, Union
+
+from dsp.modules.lm import LM
+
+## Utility functions to load models
+
+
+def load_tensorrt_model(
+    engine_dir: Union[str, Path],
+    use_py_session: Optional[bool] = False,
+    **kwargs,
+) -> tuple[Any, dict]:
+    import tensorrt_llm
+    from tensorrt_llm.runtime import ModelRunner, ModelRunnerCpp
+
+    runtime_rank = tensorrt_llm.mpi_rank()
+    runner_cls = ModelRunner if use_py_session else ModelRunnerCpp
+    runner_kwargs = {
+        "engine_dir": engine_dir,
+        "lora_dir": kwargs.get("lora_dir", None),
+        "rank": runtime_rank,
+        "lora_ckpt_source": kwargs.get("lora_ckpt_source", "hf"),
+    }
+
+    if not use_py_session:
+        engine_cpp_kwargs = {}
+        defaults = {
+            "max_batch_size": 1,
+            "max_input_len": 1024,
+            "max_output_len": 1024,
+            "max_beam_width": 1,
+            "max_attention_window_size": None,
+            "sink_token_length": None,
+        }
+
+        for key, value in defaults.items():
+            engine_cpp_kwargs[key] = kwargs.get(key, value)
+        runner_kwargs.update(**engine_cpp_kwargs)
+
+    runner = runner_cls.from_dir(**runner_kwargs)
+    return runner, runner_kwargs
+
+
+def tokenize(prompt: Union[list[dict], str], tokenizer: Any, **kwargs) -> list[int]:
```

Comment:
should this return type be list[list[int]]?

Reply:
Uhmm, so I assumed that in dspy, `basic_request` we only use a single prompt string, so I did this for only that case, where we can pass only a single string or a single conversation which is like this:

```
[{"role": "user", "content":"hello"}, {"role": "assistant", "content":"hi"}]
```